### PR TITLE
Use a proper deque to improve efficiency.

### PR DESCRIPTION
**Pull request by favonia**
This pull request is tracking the progress on solving #12 by using a better deque.

PS: bugs in the original implementation were found while changing the underlying data structures.

### CODE REVIEW

**Comment by favonia** [comment]
@jonsterling I might have found a bug... did `dropUntil` forget a `List.rev` in the current code?

**Comment by jonsterling** [comment]
> I might have found a bug.

wow! That is bad if so... It's been honestly a long time since I looked at this code, and I would not be surprised if there are some bugs in it—though I would expect a bug like this would mess up a lot of things about RedPRL, so it's a little surprising.

Maybe the easiest thing to do is add a test to https://github.com/RedPRL/sml-telescopes/blob/master/test/test.sml.

**Comment by favonia** [comment]
@jonsterling I actually did, but not pushed it yet.

**Comment by favonia** [comment]
@jonsterling The current `eq` does not care about the ordering so the testing cannot tell (without black magic). Should `subtelescope` check the ordering too?

**Comment by jonsterling** [comment]
@favonia Ahh, that's a good point. It's a little tricky, since we do want it to care about order, but `subtelescope` should permit items to be omitted.

You should feel free to make any changes to this library that you think are justified. I threw it together quite quickly, and it is probably crap. 😉

File: test/test.sml [hunk 1]
```diff
@@ -20,14 +20,19 @@ struct
...
+      val tele21 = empty >: ("2",2) >: ("1",1)
+      val tele34 = empty >: ("3",3) >: ("4",4)
     in
-      assert "refl" @@ eq op= (tele, tele);
-      assert "foldr" @@ T.foldr (fn (_, x, r) => x :: r) [] tele = [1,2,3,4];
-      assert "foldl" @@ T.foldl (fn (_, x, r) => x :: r) [] tele = [4,3,2,1];
-      assert "truncateFrom" @@ eq op= (truncateFrom tele "3", tele');
-      assert "truncateFrom/not-a-key" @@ eq op= (truncateFrom tele "not-a-key", tele)
+      assert "refl" @@ eq op= (tele1234, tele1234);
+      assert "eq/reorder" @@ not (eq op= (tele12, tele21));
```
Comment (favonia) [inline test/test.sml:29]: The current implementation fails this.

Comment (jonsterling) [reply test/test.sml:29]: good job! now we can fix it.

**Review by jonsterling — APPROVED** [review]
looks good! And I tested this in RedPRL and it does improve performance somewhat already (for `connection.prl`, looks like about a 30% speedup).

**Comment by favonia** [comment]
I also observed a 30-35% speedup. Here is the profiling:
```
0.24 seconds of CPU time (0.13 seconds GC)
       function          cur 
----------------------- -----
<gc>                    35.1%
Sequence.append         10.8%
Susp.delay               8.1%
ListDeque.showcons       8.1%
Abt.infer                5.4%
<unknown>                2.7%
TermPrinter.varBinding   2.7%
Susp.force               2.7%
Abt.mapAbs_              2.7%
RedPrlOperator.mapSym    2.7%
FreshSymbols.freshSyms   2.7%
Syntax.out               2.7%
ListDeque.append         2.7%
SplayRDict.map           2.7%
AbtSymbol.Ord.compare    2.7%
SplayTree.findAndSplay   2.7%
Abt.imprisonVariable.fn  2.7%
```

**Comment by jonsterling** [comment]
great job, @favonia! Thanks.

**Pull request merged**